criu/pie/restorer.c

```diff
@@ -1528,11 +1549,43 @@ long __export_restore_task(struct task_restore_args *args)
 	for (i = 0; i < args->vma_ios_n; i++) {
 		struct iovec *iovs = rio->iovs;
 		int nr = rio->nr_iovs;
-		ssize_t r;
+		ssize_t r = 0;
 
 		while (nr) {
-			pr_debug("Preadv %lx:%d... (%d iovs)\n", (unsigned long)iovs->iov_base, (int)iovs->iov_len, nr);
-			r = sys_preadv(args->vma_ios_fd, iovs, nr, rio->off);
+			/* Using mmap to load content as an on-demand way rather than egarly read all the content */
+			local_offset = rio->off;
+			for (j = 0; j < nr; j++) {
+				void * p = (void *)iovs[j].iov_base;
+				unsigned long end = (unsigned long)p + iovs[j].iov_len;
+				pr_debug("Change preadv to map of iov buffer %d, iov base %lx\n", j, (unsigned long)iovs[j].iov_base);
+				/* Considering the iov buffer may cross multiple regions */
+				while ((unsigned long)p < end) {
+					/* Find the vma_entry */
+					tmp_entry = lookup(p, args);
+					io_size = tmp_entry->end < end ? tmp_entry->end : end;
+					/* The length that we exactly want to map */
+					io_size -= (long)p;
+
+					/* 
+						* Now check whether it can be mapped, notice that some memory areas such as `vsyscall`,
+						* which is not VMA_AREA_REGULAR, and stack area, whose flag has MAP_GROWSDOWN still need
+						* to use `sys_pread`.
+						*/
+					if (!vma_entry_is(tmp_entry, VMA_AREA_REGULAR) || (tmp_entry->flags & MAP_GROWSDOWN) || (tmp_entry->flags & MAP_ANONYMOUS)) {
+						local_r = sys_pread(args->vma_ios_fd, p, io_size, local_offset);
+						pr_debug("Can't use mmap, still choose pread, %lx, %ld, %d\n", (unsigned long)p, io_size, args->vma_ios_fd);
+
+					} else {
+						sys_munmap(p, io_size);
+						sys_mmap(p, io_size, tmp_entry->prot | PROT_WRITE, (tmp_entry->flags & ~MAP_ANONYMOUS) | MAP_FIXED | MAP_PRIVATE, args->vma_ios_fd, local_offset);
```

Comment:
you have to check a return value

Reply:
Yes, I should check whether the return value equals to p or not. I need to check the return value of `sys_munmap` too.